edb/pgsql/compiler/relgen.py

```diff
@@ -1176,6 +1220,25 @@ def process_set_as_subquery(
                 relctx.include_rvar(
                     stmt, subrvar, ir_source.path_id, ctx=ctx)
 
+        # If we are looking at a materialized computable, running
+        # get_set_rvar on the source above may have made it show
+        # up. So try to lookup the rvar again, and if we find it,
+        # skip compiling the computable.
+        if ir_source and (new_rvar := _lookup_set_rvar(ir_set, ctx=ctx)):
+            if semi_join:
+                # We need to use DISTINCT, instead of doing an actual
+                # semi-join, unfortunately
```

Comment:
why?

Reply:
Because semi-joins work by taking a fresh, unique copy of the set (from a new base rel) and filtering it down via semi-join against the subquery. But we need to exfiltrate actual data out of the subquery, not just whether something is present.

Reply:
Hm, I think maybe we could do something like:
```
SELECT ...
    FROM <new base relation>
    CROSS JOIN LATERAL (
        SELECT ... FROM <the subquery with the rvar unpacking>
        WHERE <join condition>
        LIMIT 1
    )
```
Do you think that would be better?